### DISCUSSION TITLE
Wild idea: Integrate DuckDB to enable fast OLAP queries

### DISCUSSION THREAD

**Discussion by loraderon** [ravendb/ravendb#18251] [Ideas]
I recently found out about [DuckDB](https://duckdb.org/) beucase of the [.NET bindings](https://github.com/Giorgi/DuckDB.NET).
_DuckDB is a fast in-process analytical database_

RavenDBs could host the DuckDB library/api and expose a limited set of the capabilities for querying data by some sort of "Internal OLAP Task".

**Comment by ayende** [comment]
What sort of operations do you expect to be able to do with that?

**Reply by loraderon** [reply]
Analytical queries with grouping, aggregations & ordering.
```sql
SELECT media, sum(impressions)
FROM statistics
WHERE 
 campaignid = '329bb91c-b126-4e44-986a-9e6934221d0a'
AND
 date BETWEEN '2023-01-01' AND '2023-10-01'
GROUP BY media
ORDER BY sum(impressions) DESC
LIMIT 10
```

I really struggled on how to store & create indexes for statistics such as how many impressions/clicks per day (https://github.com/ravendb/ravendb/discussions/17488). I was able to get it to work (and it's actually pretty fast) but I have to create very specific indexes for every case.

I know that the best solution would probably be to introduce another analytical database in the system but that would introduce a lot more complexity into the system.

**Reply by ayende** [reply]
The problem is that this would require us to move the data to DuckDB, which is basiaclly just another indexing scheme.
The indexing engine is already really good here, as well as being able to run those queries natively when you are using time series. 
Have you tried that?

**Reply by loraderon** [reply]
I see no possibility to use timesseries to store statistics. There seems to be no way if indicating the sources of the stats such as device or media. So there is no possibility to group and aggregate the data on anything else than time.

I have millions of stats such as:
```json
{
    "Date": "2023-07-20",
    "CampaignId": "329bb91c-b126-4e44-986a-9e6934221d0a",
    "Media": "example.com",
    "Device": "Mobile",
    "Impressions": 21,
    "Clicks": 3,
}
```

And one of the questions I would like to have answered is "What are the top 10 medias with the most impressions for these campaigns in the given time period":

```sql
SELECT media, sum(impressions)
FROM statistics
WHERE 
 campaignid = '329bb91c-b126-4e44-986a-9e6934221d0a'
AND
 date BETWEEN '2023-01-01' AND '2023-10-01'
GROUP BY media
ORDER BY sum(impressions) DESC
LIMIT 10
```

If this is possible to accomplish with timeseries it would be great! 

At the moment I have to resort to a index that is hacky since it's not guaranteed to contain the correct values since I'm sorting and limiting the amount of medias that are reduced for each day. I see no other way to only get the medias I'm interested in. There are hundreds, sometimes even thousands, of different media for each day and campaign.

```cs
public class Statistics_TopMedia
    : AbstractIndexCreationTask<Statistics, Statistics_TopMedia.Result>
{
    public Statistics_TopMedia()
    {
        Map = stats =>
            from s in stats

            select new Result
            {
                CampaignId = s.CampaignId,
                Date = s.Date,
                TopMedias = new MediaStat[]
                {
                    new()
                    {
                        Media = s.Media,
                        Impressions = s.Devices.Sum(kv => kv.Value.Impressions),
                        Clicks = s.Devices.Sum(kv => kv.Value.Clicks),
                    }
                },
            };

        Reduce = results =>
            from result in results
            group result by new { result.CampaignId, result.Date, } into g
            select new Result
            {
                CampaignId = g.Key.CampaignId,
                Date = g.Key.Date,
                TopMedias = g.SelectMany(x => x.TopMedias)
                    .GroupBy(x => x.Media.ToLower())
                    .Select(mg => new MediaStat
                    {
                        Media = mg.Key,
                        Impressions = mg.Sum(y => y.Impressions),
                        Clicks = mg.Sum(y => y.Clicks),
                    })
                    .OrderByDescending(x => x.Impressions)
                    .Take(25)
                    .ToArray()
            };
    }

    public class MediaStat
    {
        public required string Media { get; set; }
        public required double Impressions { get; set; }
        public required double Clicks { get; set; }
    }

    public class Result
    {
        public required Guid CampaignId { get; set; }

        public required DateOnly Date { get; set; }

        public required MediaStat[] TopMedias { get; set; }
    }
}
```

**Reply by ayende** [reply]
I would model that as a separate document for each campaign, then a separate time series for each media, using the Tag for the device.

That gives you the most natural way to slice the data.

Note that RavenDB is mostly about OLTP and transactions, not reports and OLAP.
What you can do however, is to utilize the OLAP ETL feature and push Parquet files with the data to a reporting solution

**Reply by loraderon** [reply]
> I would model that as a separate document for each campaign, then a separate time series for each media, using the Tag for the device.

Thanks, I'll try to spike a poc with that setup!

> Note that RavenDB is mostly about OLTP and transactions, not reports and OLAP.

Yes, that's why I called it a "wild" idea. To enable RavenDB to also handle OLAP/reporting in a very explicit way.

**Reply by loraderon** [reply]
> The offspring of that partnership is pg_duckdb, an extension that embeds the DuckDB engine into the PostgreSQL database, allowing it to handle analytical workloads.

https://www.i-programmer.info/news/84-database/17611-duckdb-and-hydra-partner-to-get-duckdb-into-postgresql-.html
